tests/test_recommendations.py

```diff
@@ -0,0 +1,191 @@
+    import mock
+from pythonforandroid.recommendations import (
+    check_ndk_api,
+    check_ndk_version,
+    check_target_api,
+    read_ndk_version,
+    MAX_NDK_VERSION,
+    RECOMMENDED_NDK_VERSION,
+    RECOMMENDED_TARGET_API,
+    MIN_NDK_API,
+    MIN_NDK_VERSION,
+    NDK_DOWNLOAD_URL,
+    ARMEABI_MAX_TARGET_API,
+    MIN_TARGET_API,
+)
+from pythonforandroid.util import BuildInterruptingException
+running_in_py2 = int(py_version[0]) < 3
+
+
+class TestRecommendations(unittest.TestCase):
+    """
+    An inherited class of `unittest.TestCase`to test the module
+    :mod:`~pythonforandroid.recommendations`.
+    """
+
+    def setUp(self):
+        self.ndk_dir = "/opt/android/android-ndk"
+
+    @unittest.skipIf(running_in_py2, "`assertLogs` requires Python 3.4+")
+    @mock.patch("pythonforandroid.recommendations.read_ndk_version")
+    def test_check_ndk_version_greater_than_recommended(self, mock_read_ndk):
+        mock_read_ndk.return_value.version = [MAX_NDK_VERSION + 1, 0, 5232133]
+        with self.assertLogs(level="INFO") as cm:
+            check_ndk_version(self.ndk_dir)
+        mock_read_ndk.assert_called_once_with(self.ndk_dir)
+        self.assertEqual(
+            cm.output,
+            [
+                "INFO:p4a:[INFO]:    Found NDK version {ndk_current}".format(
+                    ndk_current=MAX_NDK_VERSION + 1
+                ),
+                "WARNING:p4a:[WARNING]:"
+                " Maximum recommended NDK version is {ndk_recommended}".format(
+                    ndk_recommended=RECOMMENDED_NDK_VERSION
+                ),
+                "WARNING:p4a:[WARNING]:"
+                " Newer NDKs may not be fully supported by p4a.",
+            ],
+        )
+
+    @mock.patch("pythonforandroid.recommendations.read_ndk_version")
+    def test_check_ndk_version_lower_than_recommended(self, mock_read_ndk):
+        mock_read_ndk.return_value.version = [MIN_NDK_VERSION - 1, 0, 5232133]
+        with self.assertRaises(BuildInterruptingException) as e:
+            check_ndk_version(self.ndk_dir)
+        self.assertEqual(
+            e.exception.args[0],
+            "Unsupported NDK version detected {ndk_current}"
+            "\n* Note: Minimum supported NDK version is {ndk_min}".format(
+                ndk_current=MIN_NDK_VERSION - 1, ndk_min=MIN_NDK_VERSION
+            ),
+        )
+        mock_read_ndk.assert_called_once_with(self.ndk_dir)
+
+    @unittest.skipIf(running_in_py2, "`assertLogs` requires Python 3.4+")
+    def test_check_ndk_version_error(self):
+        """
+        Test that a fake ndk dir give us two messages:
+            - first should be an `INFO` log
+            - second should be an `WARNING` log
+        """
+        with self.assertLogs(level="INFO") as cm:
+            check_ndk_version(self.ndk_dir)
+        self.assertEqual(
+            cm.output,
+            [
+                "INFO:p4a:[INFO]:    Could not determine NDK version, "
```

Comment:
I think all these exact output checks may be annoying to keep updated. How about storing the unformatted strings in recommendations.py, and importing them here when checking the printed output matches them? I wouldn't necessarily suggest that for most log output, but I think it makes sense for this kind of important message which is a meaningful string in itself and not a throwaway log.

More generally, it might be nice to have helper functions that can pattern match against the expected log level (info, warning, ...) and some of the string content, when the whole string isn't important. That's just a thought, not a request here.